### PR TITLE
[Multimodal] Refactor multimodal_manager and embed_input_ids to align with vLLM

### PR DESCRIPTION
**Pull request by kwang3939**
# Description

In tpu-inference, `MultiModalManager` has `mm_embed` flatten and padding logic causing error in Qwen3-VL torchax path model. This was earlier added to address the recompilation of `embed_input_ids` due to the different shapes of inputs to the function (https://github.com/vllm-project/tpu-inference/pull/972). @muskansh-google created a PR to address it through splitting the merged embeds in torchax path to unblock Qwen3-VL: https://github.com/vllm-project/tpu-inference/pull/2299.

The other way to address is similar to [PR](https://github.com/vllm-project/tpu-inference/pull/2113), this can be refactored to move jax path specific logic from `multimodal_manager` to inside the jax path code, so that `multimodal_manager` code works for both jax and torchax paths.

# Tests

Tested jax path:`python3 -m pytest -s -v tests/e2e/test_multi_modal_inference.py`
Checked with `export JAX_LOG_COMPILES=1` for jax path compilation.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/models/jax/utils/multi_modal_utils.py [hunk 1]
```diff
@@ -175,3 +175,22 @@ def merge_multimodal_embeddings(
         (input_ids == placeholder_token_id),
         multimodal_embeddings,
     )
+
+
+def prepare_jax_mm_embeds(mm_embeds: list[jax.Array] | None,
```
Comment (wdhongtw) [inline tpu_inference/models/jax/utils/multi_modal_utils.py]: nit: Maybe named `merge_mm_embeds`, or similar name that reflect the what it did. Or add one line of docstring if there is no good name, so it's easier for understanding later.

File: tpu_inference/models/jax/utils/multi_modal_utils.py [hunk 2]
```diff
@@ -175,3 +175,22 @@ def merge_multimodal_embeddings(
...
+                          target_pad_len: int) -> jax.Array | None:
+    if mm_embeds is None or len(mm_embeds) == 0:
+        return None
+
+    flattened_embeds = flatten_embeddings(mm_embeds)
+    if flattened_embeds.shape[0] == 0:
+        return None
```
Comment (wdhongtw) [inline tpu_inference/models/jax/utils/multi_modal_utils.py]: Is this check required?

If there is actually a non-empty list of mm_embed and the flatten tensor is empty, should we just return an empty tensor?

**Review by wdhongtw — COMMENTED** [review]
The overall change looks good to me. But still some questions.

--

Do we really need a separated `jitted_embed_input_ids_fn`?

If we keep using `embed_input_ids_fn`, there is no harm for torchax path.
We just doing some calculation in compilation manager that have no actual benifits.

Also, if we add `jitted_embed_input_ids_fn`, I suppose there is some change in TPUModelRunner that call different embed_input_ids in different branch.

**Comment by kwang3939** [comment]
> The overall change looks good to me. But still some questions.
> 
> --
> 
> Do we really need a separated `jitted_embed_input_ids_fn`?
> 
> If we keep using `embed_input_ids_fn`, there is no harm for torchax path. We just doing some calculation in compilation manager that have no actual benifits.
> 
> Also, if we add `jitted_embed_input_ids_fn`, I suppose there is some change in TPUModelRunner that call different embed_input_ids in different branch.

@wdhongtw Thanks for reviewing! Yeah, makes sense. I also think passing two looks quite duplicate. Just removed the jitted one.

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 3]
```diff
@@ -497,8 +497,6 @@ def embed_input_ids_func(
                         torch_mm_embeds = [torch_view(x) for x in mm_embeds]
                     else:
                         torch_mm_embeds = torch_view(mm_embeds)
-                    assert is_multimodal is not None
-                    torch_mm_embeds = torch_mm_embeds[is_multimodal]
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py:501]: why is this removed? I believe this was for truncation

Comment (kwang3939) [reply tpu_inference/models/vllm/vllm_model_wrapper.py:501]: Yes, it was for truncation. But right now the `mm_embeds` would be list while `is_multimodal` is a jax array.

**Review by lk-chen — COMMENTED** [review]
LGTM, would like @wdhongtw to take another look

**Review by wdhongtw — APPROVED** [review]
Have a look again and test locally.

It looks good and I confirmed that this change is not breaking `google/gemma-4-31B-it`, and it unblocks `Qwen/Qwen3-VL-2B-Thinking`.
